File: example/ups.c
#include <dfups.h>
#include <string.h>
#include <stdio.h>
#include <unistd.h>
#define ID      0x18
#define wDog    0x01<<4
#define led     0x01<<3
#define rgb     0x01<<2
#define shdown  0x01<<1
#define powam   0x01<<0

void getHelp(int fd){

    printf("===========================================^-^=======================================\n");
    printf("ups    [arg]\n");
    printf("    -i\n");
    printf("        get the information of UPS,for example:ID ,VERSION ,ENERGY ,RGB...\n");
    printf("\n");
    printf("    -rgb [numR]  [numG]  [numB]\n");
    printf("        need 5 arguments to set RGB,the <numR> <numG> <numB> range 0~255\n");
    printf("\n");
    printf("    -s   <min>\n");
    printf("        shutdown after <min>,and UPS turn off after,if no <min> will shutdown now\n");
    printf("\n");
    printf("    -t   [min]\n");
    printf("        after shutdown, the pi will be switched on after <min> minutes\n");
    printf("\n");
    printf("    -r   [min]  [exe]\n");
    printf("        pi will turn on in <min> minutes after shutdown,run <exe>\n");
    printf("\n");
    printf("    -a   [value]\n");
    printf("        Audio enabled, and 1 is turn on, 0 is turn off\n");
    printf("\n");
    printf("    -e   [value]\n");
    printf("                Set the current. When the current is less than the set value(mA),\n");
    printf("                UPS will shut down and give out power to the outside\n");
    printf("===========================================^-^=======================================\n");

}

void shutDown(int fd,int min){
//    int flag = getFunction(fd);
//    flag |= shdown;
    char buf[32]={0};
    if(min == 0){
//        setFunction(fd,flag);
        system("sudo shutdown now");
    }
    else if(min > 0){
        min = min*60;
        sleep(min);
//        setFunction(fd,flag);
        system("sudo shutdown now");
    }

}

void timer(int fd,int min){
    setTimer(fd,min);
}

void autorun(int fd,int min,char *file){
    char buf[128] = {0};
    timer(fd,min);

    FILE *fp1 = fopen("/dfups/tmp/startupconfig.txt","r");
    FILE *fp2 = fopen("./rc.local","w");
    if(fp1 == NULL){
        return;
    }
    if(fp2 == NULL){
        return;
    }


    while(fgets(buf,128,fp1)!=NULL){
        fputs(buf,fp2);
        if(strncmp(buf,"/dfups/ups_wdog",14) == 0){
            memset(buf,0,sizeof(buf));
            sprintf(buf,"%s start &\n",file);
            fputs(buf,fp2);
        }
        memset(buf,0,sizeof(buf));
    }
    fclose(fp1);
    fclose(fp2);
    system("sudo chmod 777 rc.local");
    system("sudo chgrp root rc.local");
    system("sudo chown root rc.local");
    system("sudo mv rc.local /etc/");

}

void getmsg(int fd){
        int version = getVersion(fd);
        int id = getId(fd);
        printf("id     :%X\n",id);
        printf("version:%x\n",version);
       
        int energy = getEnergy(fd);
        printf("energy :%d%\n",energy);
/*      if(energy < 6){
        printf("energy level:%d\n",energy);
        }
        else if(energy == 6){
            printf("energy :is charging\n");
        }
        else if(energy == 7){
            printf("energy :fully charge\n");
        }
*/

        
        int adc1 = getAdc1(fd);
        int adc2 = getAdc2(fd);
        printf("adc1   :%d\n",adc1);
        printf("adc2   :%d\n",adc2);
        
        int vcc = getVcc(fd);
        printf("vcc    :%d\n",vcc);

        int rgbr = getRgbR(fd);
        int rgbg = getRgbG(fd);
        int rgbb = getRgbB(fd);
        printf("rgb_r  :%d\n",rgbr);
        printf("rgb_g  :%d\n",rgbg);
        printf("rgb_b  :%d\n",rgbb);
        
        int flag = getFunction(fd);
        if(flag == 0){
            printf("Audio  :unable\n");
        }
        else if (flag == 1){
            printf("Audio  :enable\n");
        }

        int elec = getElectricity(fd);
        printf("electricity require :%d\n",elec);
}

void setrgb(int fd,int rNum,int gNum,int bNum){
//    int flag = getFunction(fd); //enable rgb
//    flag |= rgb;
//    setFunction(fd,flag);
    setRgbR(fd,rNum);     //the num in range 0~255
    setRgbG(fd,gNum);
    setRgbB(fd,bNum);
}

void setAudio(int fd,int value){
    setFunction(fd,value);
}

int main(int argc,char *argv[]){
    char buf[32]={0};
    if(argc < 2){
        printf("please input arguments, you can input:./ups -h\n");
        return -1;
    }
    int fd = Init(ID);
    
    if(strcmp(argv[1],"-s") == 0){
        int min;
        if(argc == 2){
            min = 0;
        }
        else if(argc == 3){
            min = atoi(argv[2]);
        }
        shutDown(fd,min);
    }
    else if(strcmp(argv[1],"-h") == 0){
        getHelp(fd);
    }
    else if(strcmp(argv[1],"-i") == 0){
        getmsg(fd);
    }
    else if(strcmp(argv[1],"-rgb") == 0){
        if(argc != 5){
            printf("please input arguments, you can input:./ups -h\n");
            return -1;
        }
        int rNum = atoi(argv[2]);
        int gNum = atoi(argv[3]);
        int bNum = atoi(argv[4]);
        setrgb(fd,rNum,gNum,bNum);
    }
    else if(strcmp(argv[1],"-t") == 0){
        if(argc != 3){
            printf("please input arguments, you can input:./ups -h\n");
            return -1;
        }
        int min = atoi(argv[2]);
        timer(fd,min);
    }
    else if(strcmp(argv[1],"-r") == 0){
        if(argc != 4){
            printf("please input arguments, you can input:./ups -h\n");
            return -1;
        }
        int min = atoi(argv[2]);
        autorun(fd,min,argv[3]);
    }
    else if(strcmp(argv[1],"-a") == 0){
        if(argc != 3){
            printf("please input arguments, you can input:./ups -h\n");
            return -1;
        }
        setAudio(fd,atoi(argv[2]));
    }
    else if(strcmp(argv[1],"-e") == 0){
        if(argc != 3){
            printf("please input arguments, you can input:./ups -h\n");
            return -1;
        }
        setElectricity(fd,atoi(argv[2]));
    }
    else{
        printf("please input correct arguments, you can input:./ups -h\n");
        return -1;
    }
    
}

File: pythonups/ups.py
import  _dfups as ups 
import sys
import os
import time

ID   = 0x18
wDog = 0x01<<4
led  = 0x01<<3
rgb  = 0x01<<2
shDown=0x01<<1
powAm= 0x01<<0

def shutDown(fd,min=0):
    if min==0:
        os.system("sudo shutdown now")
    elif min > 0:
        min = min*60
        time.sleep(min)
        os.system("sudo shutdown now")


def getHelp():
    print("===========================================^-^=======================================")
    print("ups     [arg]")
    print("        -i")
    print("                get the information of UPS,for example:ID ,VERSION ,ENERGY ,RGB...")
    print(" ")
    print("        -rgb [numR]  [numG]  [numB]")
    print("                need 5 arguments to set RGB,the <numR> <numG> <numB> range 0~255")
    print(" ")
    print("        -s   <min>")
    print("                shutdown pi,and UPS will turn off after")
    print(" ")
    print("        -t   [min]")
    print("                after shutdown, the pi will be switched on after <min> minutes")
    print(" ")
    print("        -r   [min]  [exe]")
    print("                pi will turn on in <min> minutes after shutdown,run <exe>")
    print(" ")
    print("        -a   [value]")
    print("                Audio enabled, and 1 is turn on, 0 is turn off")
    print(" ")
    print("        -e   [value]")
    print("                Set the current. When the current is less than the set value(mA),")
    print("                UPS will shut down and give out power to the outside")
    print("===========================================^-^=======================================")

def timer(fd,num):
    ups.setTimer(fd,num)

def autorun(fd,num,filename):
    buf = bytearray(128)
    
    os.system("sudo cp /dfups/tmp/startupconfig.txt /home/pi")
    os.system("sudo chgrp pi /home/pi/startupconfig.txt");
    os.system("sudo chown pi /home/pi/startupconfig.txt");

    fp1 = open("/home/pi/startupconfig.txt","r+")
    fp2 = open("rc.local","w+")
    
    timer(fd,num)
    while True:
        buf = fp1.readline()
        fp2.write(buf)
        fp2.flush()
        if buf[:15] == "/dfups/ups_wdog":
            print("setup success!")
            buf = filename + " start &\n"
            fp2.write(buf)
            fp2.flush()
        if not buf:
            break
    
    fp1.close()
    fp2.close()
    os.system("sudo chmod 777 rc.local")
    os.system("sudo chgrp root rc.local")
    os.system("sudo chown root rc.local")
    os.system("sudo rm /home/pi/startupconfig.txt")
    os.system("sudo mv rc.local /etc/")


def getMsg(fd):
    print("id     : %s"% hex(ups.getId(fd)))
    print("version: %s"% hex(ups.getVersion(fd)))
#    en = (int)(ups.getEnergy(fd))
#    if en < 6:
#        print("energy level:%s"% int(en))
#    elif en == 6:
#        print("energy :is charging")
#    elif en == 7:
#        print("energy :fully charge")

    print("energy : %s"% int(ups.getEnergy(fd))+"%")
    print("adc1   : %s"% hex(ups.getAdc1(fd)))
    print("adc2   : %s"% hex(ups.getAdc2(fd)))
    print("vcc    : %s"% hex(ups.getVcc(fd)))
    flag = ups.getFunction(fd)
    if flag == 1:
        print("Audio  : enable")
    elif flag == 0:
        print("Audio  : unable")
    print("R      : %s"% hex(ups.getRgbR(fd)))
    print("G      : %s"% hex(ups.getRgbG(fd)))
    print("B      : %s"% hex(ups.getRgbB(fd)))
    print("electricity require : %s"% int(ups.getElectricity(fd)))

def setrgb(fd,r,g,b):
    ups.setRgbR(fd,r)
    ups.setRgbG(fd,g)
    ups.setRgbB(fd,b)



def main():
    argc = len(sys.argv)
    if argc < 2:
        print("please input arguments,you can input:ups.py -h")
        return 

    fd = ups.Init(ID)
    if sys.argv[1] == '-s':
        min = 0
        if argc == 3:
            min = int(sys.argv[2])
            if min < 0:
                print("please input a num bigger than 0")
                return -1
        shutDown(fd,min)

    elif sys.argv[1] == '-h':
        getHelp()

    elif sys.argv[1] == '-i':
        getMsg(fd)

    elif sys.argv[1] == '-rgb':
        if(argc < 5):
            print("please input arguments,you can input:ups.py -h")
            return 

        rNum = int(sys.argv[2])
        gNum = int(sys.argv[3])
        bNum = int(sys.argv[4])
        setrgb(fd,rNum,gNum,bNum)

    elif sys.argv[1] == '-t':
        if argc<3:
            print("please input arguments,you can input:ups.py -h")
            return
        num = int(sys.argv[2])
        timer(fd,num)

    elif sys.argv[1] == '-r':
        if argc<4:
            print("please input arguments,you can input:ups.py -h")
            return 
        num = int(sys.argv[2])
        autorun(fd,num,sys.argv[3])
    
    elif sys.argv[1] == '-e':
        if argc<3:
            print("please input arguments,you can input:ups.py -h")
            return
        num = int(sys.argv[2])
        ups.setElectricity(fd,num)
        
    elif sys.argv[1] == '-a':
        if argc<3:
            print("please input arguments,you can input:ups.py -h")
            return
        num = int(sys.argv[2])
        ups.setFunction(fd,num)
    
    
    else :
        print("please input arguments,you can input:ups.py -h")

    
main()


